docs/proposals/04-etcd-member-custom-resource.md

```diff
@@ -0,0 +1,561 @@
+
+While `State` is sufficient for many human operators, the notion of a `SubState` provides operators with an insight about the discrete stage of an etcd-member in its lifecycle. For example, consider a top-level `State: Starting`, which indicates that an etcd-member is starting. `Starting` is meant to be a transient state for an etcd-member. If an etcd-member remains in this `State` longer than expected, then an operator would require additional insight, which the authors propose to provide via `SubState` (in this case, the possible `SubStates` could be `PendingLearner` and `Learner`, which are detailed in the following sections).
+
+At present, these states are not captured and only the final state is known - i.e the etcd-member either fails to come up (all re-attempts to bring up the pod via the StatefulSet controller has exhausted) or it comes up. Getting an insight into all its state transitions would help in diagnostics.
+
+The status of an etcd-member at any given point in time can be best categorized as a combination of a top-level `State` and a `SubState`. The authors propose to introduce the following states and sub-states:
+
+#### States and Sub-States
+
+> **NOTE**: Abbreviations have been used wherever possible, only to represent sub-states. These representations are chosen only for brevity and will have proper longer names.
+
+| States        | Sub-States                 | Description                                                  |
+| ------------- | -------------------------- | ------------------------------------------------------------ |
+| New           | -                          | Every newly created etcd-member will start in this state and is termed as the initial state or the start state. |
+| Initializing  | DBV-S (DBValidationSanity) | This state denotes that backup-restore container in etcd-member pod has started initialization. Sub-State `DBV-S` which is an abbreviation for `DBValidationSanity` denotes that currently sanity etcd DB validation is in progress. |
+| Initializing  | DBV-F (DBValidationFull)   | This state denotes that backup-restore container in etcd-member pod has started initialization. Sub-State `DBV-F` which is an abbreviation for `DBValidationFull` denotes that currently full etcd DB validation is in progress. |
+| Initializing  | R (Restoration)            | This state denotes that backup-restore container in etcd-member pod has started initialization. Sub-State `R` which is an abbreviation for `Restoration` denotes that DB validation failed and now backup-restore has commenced restoration of etcd DB from the backup (comprising of full snapshot and delta-snapshots). An etcd-member will transition to this sub-state only when it is part of a single-node etcd-cluster. |
```

Comment:
I think, we are classifying too many states and sub-states. We should be very minimalistic here, otherwise backup-restore needs to connect to the Kubernetes client for too many times. This, in turn, would widen the scope of failures.

Reply:
The total number of states are 4 which is quite small number. Coming to sub-states, during initialisation a member will only enter into 1 state (DB validation) and Restoration will only be when there is just 1 member (single node etcd cluster). Coming to starting -> there are a fixed 2 states (only when the member is going to be added as a learner), there are cases when it directly jumps to `Started` state. In voting member state there are 2 states and any member which finishes learning always goes to just `Follower` state and only upon leader election changes the state would change.

Just like in any k8s - all well-written controllers are level-triggered and not edge-triggered. Therefore entities which are changing the status to which a controller reacts need to be eventually consistent only. I do not see an issue at this point.

Reply:
Okay. I was worried about the amount of network traffics. Moreover as we are critically depending on these state changes, I was afraid that there might be many edge cases. The edge cases are combination of start-restart-delay in volume attachment-temporary network failures. You can include the edge cases in this doc itself and we can contribute if there are any edge cases yet to be added.

Reply:
The state changes are not going to be that frequent and therefore network traffic should not be a worry IMHO.
As explained above druid will never react based on edges. In any distributed environment there is always a possibility of late update or failure to update. Reconcilers will always be designed to ensure that deal with these situations.

>You can include the edge cases in this doc itself and we can contribute if there are any edge cases yet to be added.

Since you pointed out that there are edge case then may be it would be better that you list at least a few down as i am sure you have already thought through this a bit and its fresh in your mind.

Reply:
As discussed during the meeting on Wed 23 Aug:
We agreed that during the implementation care should be taken to handle the following:
- Failure or delay to record state changes from backup-sidecar should be handled by druid.
- Keep an eye on cases where repeated restarts are happening causing frequent change of state/sub-state. Re-look at the current handling where containers are repeatedly restarted.